### DISCUSSION TITLE
Not sending events to Facebook on iOS 14+

### DISCUSSION THREAD

**Discussion by andreshndz** [oddbit/flutter_facebook_app_events#122] [Q&A] [answered]
## Expected Behavior
Sending events from iOS 14 and receive them at the events tester.

## Current Behavior
No events received.

## Possible Solution
I read about upgrade Facebook SDK version at least to 8.0 (9.0 if you need login), but I think that is not the problem. Here we use 9.1

## Steps to Reproduce (for bugs)
1. Send an event
2. On IntelliJ logs, you can see everything is working well and it was sent
3. Nothing appears at the events tester.

## Context
We are working to improve our analytics for better campaigns and better KPIs.

## Your Environment
* Operating System and version: MacOS Big Sur
* flutter_facebook_app_events version:  0.11.2

**Comment by tonishiva** [comment]
I have the same problem. Please let me know if you find any solution.

**Comment by faithomotoso** [comment]
I've been having the same issue. I got it to work somehow by initializing in `AppDelegate.swift` and adding `Settings.setAdvertiserTrackingEnabled(true)` after initializing but it stopped working after a while.

**Comment by faithomotoso** [comment]
Been able to resolve this. You have to request tracking permission before logging events would work. I used this package https://pub.dev/packages/app_tracking_transparency.

**Comment by jpiabrantes** [comment]
I am having the same problem. Do I need to update to the 0.12 version? I am also using 0.11.2 to avoid the null safety..

**Comment by faithomotoso** [comment]
@jpiabrantes I don't think you need to update. You have to request a permission from the user before event tracking would work on iOS 14.5

**Comment by jpiabrantes** [comment]
@faithomotoso I am asking for permission with https://pub.dev/packages/app_tracking_transparency and it is still not working. I really don't want to update to 2.0.

**Comment by faithomotoso** [comment]
@jpiabrantes go to settings and select the app you're working on, check if there's an allow tracking button. Also it takes a while for events to be sent to facebook, about 20 minutes or so

**Comment by jpiabrantes** [comment]
@faithomotoso There's an allow tracking button, and it's on! I am testing the facebook events as it is explained [here](https://business.facebook.com/business/help/2040882565969969?id=1205376682832142) they say events should appear within 30 secs.

**Comment by faithomotoso** [comment]
@jpiabrantes Okay, that part never worked for me, probably because of SKAdAPI on iOS 14.5. But wait a while, should show up in the main events.

**Comment by jpiabrantes** [comment]
@faithomotoso One event is getting through (the View Content), however, it does not show the parameters I've sent. The other events don't even appear on the list. Any ideas?

**Comment by faithomotoso** [comment]
@jpiabrantes Confirm if the time for the events is `x minutes ago` if it is, it'll show up. SKAdAPI delays the delivery of events to Facebook.

**Comment by jpiabrantes** [comment]
@faithomotoso I've just tried the real-time events test with Android. No events are shown too.

**Comment by faithomotoso** [comment]
@jpiabrantes Do you mind sharing how you implemented it?

**Comment by AntonLT** [comment]
@faithomotoso Automatically logged events are working in your setup?

**Comment by faithomotoso** [comment]
@AntonLT yes. App installs and launch gets logged. Although I've only ever used a simulator.

**Comment by AntonLT** [comment]
@faithomotoso Please, describe what you have done so it's working

**Comment by faithomotoso** [comment]
@AntonLT https://github.com/oddbit/flutter_facebook_app_events/issues/113#issuecomment-875695491

**Comment by AntonLT** [comment]
@faithomotoso Thank you! I meant what steps you have done prior to the moment of requesting tracking permission. It would help me much.

**Comment by faithomotoso** [comment]
@AntonLT I followed the docs to setup, added the required keys in `info.plist` and initialized in `AppDelegate.swift` with 
```
ApplicationDelegate.shared.application(application, didFinishLaunchingWithOptions: launchOptions)
Settings.setAdvertiserTrackingEnabled(true) // Not sure if this is required
```
remember to import `FBSDKCoreKit` in `AppDelegate.swift`

**Comment by AntonLT** [comment]
@faithomotoso Thank you! What have you entered in CLIENT-TOKEN? Did you get the token somewhere? Maybe you could provide some pseudo-example?

**Comment by faithomotoso** [comment]
@AntonLT yes I did. Forgot about that. You'll find it under advanced settings.

**Comment by jpiabrantes** [comment]
> @AntonLT yes I did. Forgot about that. You'll find it under advanced settings.

This solved my problem! YAAAAY 
Thank you!


However, I made no changes to my `AppDelegate.swift`.

**Comment by faithomotoso** [comment]
@jpiabrantes Good to hear! Yeah was trying a lot of things with `AppDelegate.swift`, not sure if it was needed or not.

**Comment by AntonLT** [comment]
Oh, I have done everything, but automatically logged events don't work and only for ios...

**Comment by faithomotoso** [comment]
@AntonLT Please confirm if there's an `Enable tracking` options under settings for your app in iOS

**Comment by AntonLT** [comment]
@faithomotoso, yeah, there are 😔

**Comment by faithomotoso** [comment, marked as answer]
@AntonLT in `info.plist` add 
```
<key>FacebookAutoLogAppEventsEnabled</key>
<true/>
```

**Comment by AntonLT** [comment]
OMG, it's working! Thank you so much, @faithomotoso.

**Comment by jpiabrantes** [comment]
> @AntonLT in `info.plist` add
> 
> ```
> <key>FacebookAutoLogAppEventsEnabled</key>
> <true/>
> ```

I don't have this on my `info.plist` and I am getting the automatic logged events. There is a lot of art in getting this to work 😂

**Comment by hahai96** [comment]
@jpiabrantes @faithomotoso What version are you using? Events are not logged on IOS, even though I have fully configured.

**Comment by hahai96** [comment]
![image](https://user-images.githubusercontent.com/23280338/126465945-bec23f92-6978-40a2-8e58-433b0609800d.png)

**Comment by jpiabrantes** [comment]
I am on 0.11.2.

I don't have the LSApplicationQueriesSchemes.

**Comment by hahai96** [comment]
![image](https://user-images.githubusercontent.com/23280338/126477077-a32c10b5-ef26-4080-9f6c-3bc408ada6ef.png)

**Comment by hahai96** [comment]
@jpiabrantes Do you need any special configuration?

**Comment by faithomotoso** [comment]
@hahai96 Can you check under settings if your app has the `Enable tracking` option?

**Comment by armandojimenez** [comment]
Any of the stuff mentioned here does not work for me

**Comment by jenkliu** [comment]
I added the app_tracking_transparency library and made sure I have "Enable tracking" on in Settings. 

I am getting some Purchase events in my Facebook ads dashboard, however my analytics dashboard shows 5-6x more Purchase events than that.

Is this because a user needs to tap "Allow tracking" on the dialogue in order for any event of theirs to be logged?  

I read [here](https://commonthreadco.com/blogs/coachs-corner/facebook-ads-ios-14-changes-affect-ecommerce) that Facebook will still track the top-priority event (Purchases) for users who opt out of tracking, but am not sure if I'm missing something.

@jpiabrantes @faithomotoso @AntonLT are you tracking Purchase events? If so, can you clarify if you are receiving ALL Purchase events or only those for users who explicitly opt in to tracking? Thank you!

**Comment by faithomotoso** [comment]
@jenkliu I'm not tracking purchase events. I think users have to allow tracking before anything is sent to facebook.

**Comment by DennisAlund** [comment]
I will convert this to a discussion, as it seems to be a very relevant discussion with good information for future troubleshooting. But it doesn't seem to be an issue to be fixed in the plugin, as I understand from the thread.

**Comment by mayouf** [comment]
Hi everyone,

### My  Environment
**Operating System and version:** MacOS Big Sur
**flutter_facebook_app_events version:** 0.13.4
**Flutter doctor -v**  _(expand below)_ :
<details><summary> see flutter doctor</summary>

```
 [✓] Flutter (Channel stable, 2.2.3, on macOS 11.6 20G165 darwin-x64, locale en-FR)
    • Flutter version 2.2.3 at /Users/bob/development/flutter
    • Framework revision f4abaa0735 (3 months ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/bob/Library/Android/sdk
    • Platform android-31, build-tools 30.0.3
    • Java binary at: /Library/Java/JavaVirtualMachines/jdk-16.0.2.jdk/Contents/Home/bin/java
    • Java version Java(TM) SE Runtime Environment (build 16.0.2+7-67)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.11.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] Connected device (2 available)
    • iPhone (mobile) • ABCDEF		 • ios            • iOS 14.6
    • Chrome (web)    • chrome                        • web-javascript • Google Chrome 93.0.4577.63
```
</details>

### ISSUE DESCRIPTION

I do receive all events on Android and IOS 12 on the event real-time tester. 

BUT, I don't receive any event on IOS 14 !!!

**NOTE :** _I underline that on the same code, the IOS 12 iphone send all the events, so there is no issue on IOS 12_. 

 I tried every point mentionned in this discussion, but nothing worked for me.

So here all the fixing I tried and please let me know if something is missing here :

- I ask permission with the plugin 'https://pub.dev/packages/app_tracking_transparency'.
  The permission request is sent in the `main()` function just after the `WidgetsFlutterBinding.ensureInitialized()`.
  The goal here was to allow tracking before anything is sent.
  _(For info, I did ask the permission when home screen was built, but it didn't change anything)._

- I setup my `<key>FacebookClientToken</key>` in `info.plist`.
- I verified that option "Enable tracking" under settings in Facebook Developer.
- I verified that `<key>FacebookAutoLogAppEventsEnabled</key>` is set to `true` in `info.plist`.
- I verified that `<key>FacebookAdvertiserIDCollectionEnabled</key>` is set to `true` in `info.plist`.

**NOTE:** I even set these two keys on `main()` like below, but nothing changed, at all.
```
final facebookAppEvents = FacebookAppEvents();
void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  facebookAppEvents.setAutoLogAppEventsEnabled(true);
  facebookAppEvents.setAdvertiserTracking(enabled: true);
  ....
  }
```

- I sent a customized log event on home screen building, like below:
```
@override
Widget build(BuildContext context) {
   final facebookAppEvents = FacebookAppEvents();
   facebookAppEvents.logEvent(
      name: 'APP STARTED',
      parameters: {
        'EVENT_NAME_ACTIVATED_APP': platform,
      },
  );
```

NONE of the actions above solved the IOS 14 issue.

Please anyone has a clue of why we do not receive events on IOS 14 ?

Regards,

**Reply by faithomotoso** [reply]
@mayouf can you try 
`await facebookAppEvents.setAdvertiserTracking(enabled: true);`
before logging an event

**Reply by DennisAlund** [reply]
Here's a trick to ensure that you await all initializations before going into the app

```dart
class MyApp extends StatelessWidget {
  // You can chain more initializations here if you need more things to finish
  // before  you display the app or go into widgets that depend on initializations.
  final _appInitialization = facebookAppEvents.setAdvertiserTracking(enabled: true))
      .then((_) => someOtherInitialization())
      .then((_) => andAnotherInitialization());
      
  @override
  Widget build(BuildContext context) {
    return FutureBuilder<void>(
      future: _appInitialization,
      builder: (context, snapshot) {
        if (snapshot.connectionState != ConnectionState.done) {
          return MaterialApp(
            title: 'My loading app',
            theme: appTheme,
            home: SplashScreen(),
          );
        }

        // Application ready to be used
        return MaterialApp(
          title: 'My initialized app',
          theme: appTheme,
          home: HomeScreen(),
        );
      },
    );
  }
}
```

**Reply by mayouf** [reply]
Isn't it the same as below ?
```
class MyApp extends StatelessWidget {
    await facebookAppEvents.setAutoLogAppEventsEnabled(true);
    await facebookAppEvents.setAdvertiserTracking(enabled: true);
    .....
}
```

**Reply by shariqA7** [reply]
@mayouf how you add events for your android device?

**Comment by mayouf** [comment]
Thx for answering.

It did not change anything. 

On IOS 12.5.4 : it works.
On IOS 14.6    : NO EVENTS

just in case, here is my code, which is in main():
```
import 'package:facebook_app_events/facebook_app_events.dart';
final facebookAppEvents = FacebookAppEvents();

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await facebookAppEvents.setAutoLogAppEventsEnabled(true);
  await facebookAppEvents.setAdvertiserTracking(enabled: true);

  if(Platform.isIOS){
    try{
      String iosVersionString = Platform.operatingSystemVersion;
      iosVersionString = iosVersionString.split(".")[0];
      double iosVersion = double.parse(iosVersionString);

      if(iosVersion>=14.0){
        final status = await AppTrackingTransparency.requestTrackingAuthorization();
      }

    }on PlatformException{
      print("-----ERROR checking IOS version -----");
    }
  }
...
}
```

**Reply by faithomotoso** [reply]
Does the app tracking request work well? Under iOS settings, select your app, do you see an option to turn on/off app tracking?

**Reply by mayouf** [reply]
The app tracking dialog appears and I can click on "Authorize"...so I guess it works.

The option to turn on/off app tracking is ON. I doubled check with another iPhone with the IOS 12 version.

When I plug the IOS 12 iPhone, everything is OK.


- I even tried to:

To follow these best practice https://developers.facebook.com/docs/app-events/gdpr-compliance/

So I set all info.plist keys to false and then set all of them to true when the app tracking is Authorized.

I start the app with the keys set to false:
```
<key>FacebookAutoLogAppEventsEnabled</key>
<false/>
<key>FacebookAdvertiserIDCollectionEnabled</key>
<false/>
```

below I set the keys to true if app tracking is validated:
```
import 'package:facebook_app_events/facebook_app_events.dart';

final facebookAppEvents = FacebookAppEvents();

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  if(Platform.isIOS){
    try{
      String iosVersionString = Platform.operatingSystemVersion;
      iosVersionString = iosVersionString.split(".")[0];
      double iosVersion = double.parse(iosVersionString);

      if(iosVersion>=14.0){
        final status = await AppTrackingTransparency.requestTrackingAuthorization();
        switch(status.index){
          case 3:
            await facebookAppEvents.setAutoLogAppEventsEnabled(true);
            await facebookAppEvents.setAdvertiserTracking(enabled: true);
            break;
          case 2:
            facebookAppEvents.setAdvertiserTracking(enabled: false);
            break;
        }
      }
    }on PlatformException{
      print("-----ERROR checking IOS version -----");
    }
...
}
```

- I also tried to set this key on `info.plist`
```
<key>FacebookAutoInitEnabled</key>
<true/>
```

Nothing changed !!

**Reply by faithomotoso** [reply]
In events manager, have you set the list of events and priorities to work with SkAdNetwork?

**Reply by mayouf** [reply]
Hi @faithomotoso ,

Thx for following up. I think everything is OK, but just in case and for other people to see, I show below my Facebook Developer App Account and also my Event Manager from Business Manager.

### On FB Developper

my app events setting are shown below:

<img src="https://user-images.githubusercontent.com/5865600/133410921-6601797e-d35c-41f7-a970-c9aed316055f.png" width="600">

### On Event Manager

My events manager settings are shown below:

<img src="https://user-images.githubusercontent.com/5865600/133411828-6b62ad78-ca92-4a3e-a8bd-53cc4b08ff6e.png" width="600">
<img src="https://user-images.githubusercontent.com/5865600/133411955-94ffb29c-54c6-4cd1-881a-f86f3f5d8911.png" width="600">

**Reply by faithomotoso** [reply]
![image](https://user-images.githubusercontent.com/31479497/133414686-4a804b2a-ff52-4bfc-a4ab-390512a2a510.png)

What's under `Start Setting up Events`?

**Reply by mayouf** [reply]
It takes to this screen below. If you:

![image](https://user-images.githubusercontent.com/5865600/133416164-5f33e63d-e670-49d4-9e9b-7087a5c10032.png)

**Reply by faithomotoso** [reply]
Yeah this section has to be setup to work with iOS 14+. Try to log more events on iOS12 - 13 and android for it to show up here so you'll set the priority

![image](https://user-images.githubusercontent.com/31479497/133416799-136aefd1-e36d-4f5b-9a61-bea15856786f.png)

**Reply by mayouf** [reply]
I don't find this section. Is it on event manager or FB developer ?

**Reply by faithomotoso** [reply]
Event manager, it's supposed to be here https://github.com/oddbit/flutter_facebook_app_events/discussions/122#discussioncomment-1332573 but facebook isn't seeing enough events to trigger that option

**Comment by binnicordova** [comment]
but, how use the uuid generated by appTrackingTransparencyProvider.requestTrackingAuthorization()
on here facebookAppEvents.logEvent ????

**Comment by santiLabhouse** [comment]
Hi, I did all the steps: added all in info.plist (FacebookAppID, FacebookClientToken, FacebookAdvertiserIDCollectionEnabled, CFBundleURLSchemes, etc), I request the app transparency tracking inside the app, and also upgraded FacebookSDK to latest version, but still I don't receive events in the test event console. It's strange because I can see some standard events logged (like app install or activated app) in the general info tab on facebook business manager, but others like Subscription or Purchase are not showing. And I can't test in real time because I can't see any event logging in the test event console. Someone knows why this could be happening? Thanks

**Comment by mayouf** [comment]
Hi @santiLabhouse ,

It's been I long time I had this issue, and I remember I finally received events a few days later. Even my SDK update was detected days later.

It seems like IOS14 has a mechanism in which conversion events are not real-time and are delayed. 

This mechanism is called SKAdNetwork, it is set by Apple to protect private data related to attribution. It is pretty complicated to me as the mechanism is quite technically.

If you want to learn more about it, you can read this:
[facebook-ads-delayed-reporting-after-ios14-explained](https://www.impression.co.uk/blog/facebook-ads-delayed-reporting-after-ios14-explained/)
[why-is-skadnetwork-data-delayed-what-are-the-looping-timers](https://help.branch.io/faq/docs/why-is-skadnetwork-data-delayed-what-are-the-looping-timers)

Im not sure if this explains why events are not received on the test console, but in my case, it made sense as I received test events days later. Everything worked well but only with a delayed time.

**Comment by santiLabhouse** [comment]
Hi @mayouf , thank you for your explanation. I will wait some days to see if I receive events in the test console with delayed time. I hope that the configuration is correct and the problem remains on the SKAdNetwork.

**Comment by kevinzDzul** [comment]
The same thing happened to me, I couldn't track the events of the device in ios. I solved it in the following way.
1.- activate in the facebook application "allow tracking"
in iOS setting (iOS settings > Facebook > Allow Tracking).

![File](https://user-images.githubusercontent.com/88689412/151715444-d2a0d5d7-de4d-4b30-8bbc-a367f51bff4c.jpg)


2.- restart the facebook application and leave it in the background.

3.- restart your application.

after this I was able to see my application events.
In my case I could only receive the fb_mobile_install event because it is the only one that recognizes fb events by default, but if you configured the click event you should already be able to see it in the console.


```
facebookAppEvents.logEvent(
    name: 'Test Link Clicked',
    parameters: {'url':  'www.google.com'},
);
```
In the same way, check that you have added your platform to the configuration in the IOS configuration panel.

<img width="785" alt="Captura de Pantalla 2022-01-30 a la(s) 2 01 36 p m" src="https://user-images.githubusercontent.com/88689412/151715566-74b9f2f2-5ee2-46be-9ea0-b83e649effb0.png">

As a comment in my case I had to add this library to give the application permission to track the user's information by the new apple policies.

[tracking transparency](https://pub.dev/packages/app_tracking_transparency)

I share my validation code in flutter.

```dart
import 'dart:io' show Platform;
import 'package:app_tracking_transparency/app_tracking_transparency.dart';
import 'package:facebook_app_events/facebook_app_events.dart';

class FacebookTracking {
  FacebookTracking();

  Future<void> _facebookTracking() async {
    FacebookAppEvents facebookAppEvents = FacebookAppEvents();
    await facebookAppEvents.setAdvertiserTracking(enabled: true);
    await facebookAppEvents.setAutoLogAppEventsEnabled(true);
  }

  Future<void> _iosValidation() async {
    //Request tracking to user in ios devices.
    TrackingStatus status =
        await AppTrackingTransparency.requestTrackingAuthorization();
    if (status == TrackingStatus.authorized) {
      await _facebookTracking();
    }
  }

  Future<void> _androidValidation() async {
    await _facebookTracking();
  }

  Future<void> init() async {
    if (Platform.isAndroid) {
      await _androidValidation();
    }
    if (Platform.isIOS) {
      await _iosValidation();
    }
  }
}

```

I hope this helps you, it worked for me but each problem can be different in each case.

**Reply by SJGeeKo** [reply]
You saved me! Thanks! it works like a charm.

**Reply by DennisAlund** [reply]
This is interesting. Do you have a good link to any Facebook developer docs about this. I'd like to find a way that this can make its way into README if it's either possible for people to turn off tracking or even worse, if the tracking is turned off by default.

**Reply by flrngel** [reply]
4. If the instruction above doesn't work, update your Facebook app.

**Reply by christianr11** [reply]
This works for me 👌

**Comment by xalikoutis** [comment]
People and @DennisAlund the package works as expected for IOS 12 & 14 ONLY if you have already the Facebook App installed on your device, with an already activated user.
I believe that we have to add this on README

**Reply by DennisAlund** [reply]
Interesting. So you're saying that Facebook App Events SDK, natively **also** only works for iOS devices that has Facebook App installed? 

This seems inconvenient out of Facebook's own interest 🤔 ... I'd expect that would want it to work regardless of if the FB app is also installed. Similar to how I believe that their website tracking is covering your movements across sites even if you're not logged in to Facebook with a cookie on your browser... 

Let's dig into this deeper. There seem to be many different experiences across the community that use the plugin, so you're right it would be awesome to update the README with a consensus of the requirements or any quirks.

Do you have any documentation from Facebooks SDK documentation or some reproducible test scenario that we can look at?

**Reply by xalikoutis** [reply]
Sorry for my late response, i will look into this
